### ISSUE TITLE
[Feature Request] Download links for games that install via the app

### ISSUE THREAD

**Issue by hach-que** [itchio/itch.io#935]
It'd be great if there was a way to prompt users to install a game and have that download link automatically install the itch.io app if needed. That way as developers we can just provide a download link for the game, and this link will:

1. Automatically install the itch.io app if it's not already installed
2. Tell the itch.io app to start downloading/installing the specified game

This probably needs some things to work:

* The installers need to have some config that can be parameterized that they can read to know if they should launch the itch.io app and install a game
* The itch.io app itself probably needs a URL handler (#701) or some local way for the installer to tell it to start installing a game
* The itch.io servers need to be able to provide dynamic installers on demand
  * The installer needs to have some well known template string embedded inside it of a fixed length (usually this is something like `__INSTALLER_DYNAMIC_CONFIG__` followed by underscores to pad out to maximum length). Then the itch.io servers grab the generic installer executable, and replace that string inside the executable with the JSON config as needed.
  * If the installers are currently signed (like Windows codesign), then the itch.io servers probably need to run codesign on new executable before serving it.
  * The resulting installer can be cached pretty much forever for each game after the first hit (or at least until a new version of the generic installer comes out).

**Comment by hach-que** [comment]
The template to be replaced could probably just be thrown in `data/config.json` and then retrieve at runtime since it looks like go-bindata already handles embedding assets. With a little more work, we could also change the banner image shown in the installer to use the game's thumbnail or banner image.

**Comment by hach-que** [comment]
It also looks like running the following command:

```
jrhod@DESKTOP-2P1EGQT C:\Users\jrhod\Documents\Projects\MinuteOfMayhem
$ C:\Users\jrhod\AppData\Local\itch\itch.exe "itch://install/?game_id=57522&upload_id=176489"
```

Causes the itch app to prompt for install, which kind of indicates that #701 might already be solved on the URL handler side of things.

**Comment by fasterthanlime** [comment]
I'm tending to other bugs atm but reading this thread / thinking about it in the background.

The URL handler side of things is definitely handled already - dynamically generating an itch-setup exe should be the last thing we consider imho, as CI pipelines range from 40s to 1 minute, which is imho an unacceptable waiting time to start a download.

Thinking about this, I'd rather have a solution where the itch.io servers register that you wanted to install a given game, then serves you the regular setup, and, after logging in from the app, prompt you to install that game.

**Comment by hach-que** [comment]
In this case you're not dynamically building the itch setup. You're taking the already built, generic installer and replacing a string inside the binary with the config and serving the modified file.

**Comment by fasterthanlime** [comment]
> In this case you're not dynamically building the itch setup. You're taking the already built, generic installer and replacing a string inside the binary with the config and serving the modified file

Right, I see the part about the padding now - that still involves running a service on Windows and macOS servers (for codesigning) though, mhh.

**Comment by hach-que** [comment]
Does the installer need to embed an API key? Looking at the itch app source
code, it looks like accessing a itch://install URL while logged out will
queue it up to be loaded after the user logs in.

**Comment by fasterthanlime** [comment]
> Does the installer need to embed an API key? Looking at the itch app source code, it looks like accessing a itch://install URL while logged out will queue it up to be loaded after the user logs in.

Ah, yes - "after it logs in" being the key word here. I'm ready to bet that the login flow is currently a bigger issue than anything else- seeing as you need to remember your username/password, and then potentially solve a captcha or get your 2FA device out. A magic link that signs you in automatically (if necessary) and then queues install sounds like it lifts a lot more obstacles to me!
